src/hooks/useOnResizeRef.ts

```diff
@@ -0,0 +1,64 @@
+import { useEffect, useRef } from 'react';
+import { throttle } from '../util/throttle';
+import { useLatestFn } from './useLatest';
+
+type CB = (el: Element) => any;
+
+const observedElements = new WeakMap<Element, Set<CB>>();
+
+function elementIsObserved(el: Element, cb: CB) {
+  return !!observedElements.get(el)?.has(cb);
+}
+
+function addObservedElement(el: Element, cb: CB) {
+  if (!observedElements.has(el)) {
+    observedElements.set(el, new Set([cb]));
+  }
+  observedElements.get(el)!.add(cb);
+}
+
+// This hook returns a callback ref. Here is how it's used:
+/**
+ * const onResize = (el) => {};
+ * const callbackRef = useOnResizeRef(onResize);
+ * <div ref={callbackRef} />
+ */
+// The subscription to resize events is set up in the callback ref.
+// This ensures the subscription is actually set up. If this hook tried to set up the subscription in a useEffect,
+// there's a good chance it would fail. This is because the element usually comes from a ref and starts out as null.
+// The effect would run, but the subscription might not be set up because ref.current is null.
+// ref.current changing from null to something else will not cause a re-render, so the effect would not run again.
+export function useOnResizeRef<T extends Element>(callback: (el: T) => any) {
+  const cleanupRef = useRef(() => {});
+  const cb = useLatestFn(callback) as CB;
+  const intervalCountRef = useRef(0);
+
+  useEffect(() => cleanupRef.current, []);
+
+  return (el: T | null) => {
+    if (el && !elementIsObserved(el, cb) && typeof ResizeObserver !== 'undefined') {
+      addObservedElement(el, cb);
+
+      const throttledCb = throttle(cb, 100);
+      const resizeObserver = new ResizeObserver(() => throttledCb(el));
+
+      resizeObserver.observe(el);
+
+      // ResizeObserver doesn't catch resizes due to font size changes.
+      // This is a hack to catch those resizes.
```

Comment:
`font-size` changes outside of the first 10 seconds of the element being observed are not caught. This was a frustrating issue and I almost scrapped using `ResizeObserver` in favor of only using an interval (at a faster interval than 1 second).
The callback is throttled to prevent it being called more than once in 100ms. It shouldn't matter how many times the callback is called, resizes can happen any number of times and at any time.

Reply:
Darn, that's such an annoying limitation of the observer. I think your interval is fine. I imagine it's unlikely for a user to change the font-size once they're settled on the page after that window anyways.

Ah thanks for explaining! I missed the throttle, thanks!

Reply:
I was wrong about the `ResizeObserver`. It does catch resizes due to font changes. The issue is it does not observe changes that do not affect the elements actual size. This makes sense, can be a problem:

1. The container for `children` starts exactly at `maxLines` and fits everything.
2. We check there is no overflow (`scrollHeight` same as `clientHeight`) and hide the "Show More" button.
3. A font loads, making the text larger, which adds a new line.
4. The container does not change size, it was already at `maxLines`. Yes, its `scrollHeight` changes, but `ResizeObserver` does not observe this because this is not part of the element's visible size.
5. `ResizeObserver` callback is not called and the "Show More" button remains hidden.

This is a rare edge case but also a pretty bad one if it does happen. I decided to use an interval to continuously check the `scrollHeight` instead of `ResizeObserver`. It's set to check every 200ms. Shouldn't be too bad for performance.